### ISSUE TITLE
Sharing a multi-GB library stalls MuWire

### ISSUE THREAD

**Issue by LoveIsGrief** [zlatinb/muwire#35]
I have a media library that's about pretty big (a few hundred GB). Trying to share that simply shuts down MuWire after a while. The number of connections drop to 1 and `files.json` contains nothing.

I have a feeling `files.json` is the bottleneck here. It might be better to either:

 - use a DB (sqlite is slow af, maybe there are other file DBs)
 - use a folder structure e.g :
   * `files/<filepath hash 0-5>/<filepath hash 6-31>.json`
   * `files/<filepath hash>.json`
   * `files/<share root>/<path to file>/<filename>.json`

Maybe I'm wrong.

------------------------------

Maybe it's unrelated, but I also go this error in my I2P router

> ERROR [JobQueue 2/5] .client.ClientConnectionRunner: Disconnecting the client - Too many leaseset request fails

**Comment by zlatinb** [comment]
When sharing large number of large files there are possible degenerate states but shutting down isn't one of them.  I suspect the shut down is caused by the `ERROR` you're seeing in the router logs; if MW loses the I2CP connection to the I2P router it will shut down.

The possible degenerate states are usually accompanied with very high cpu usage.  When hashing normally, MW should use up to 1 cpu core.  But if it starts to run out of memory, it will go into a garbage collection "death spiral" which will cause it to use up all available cpu cores without making any real progress.  A side effect that has been reported when MW starts to run out of memory is that the directory for temp files starts to fill up whereas `files.json` doesn't get updated.

In healthy state, MW will try to persist `files.json` once a minute.  The persisting process writes to a temporary file (usually in `/tmp`) then copies the file over `files.json` and deletes the temporary file.  I've heard reports where an `OutOfMemoryError` gets thrown while writing to the temp file and that leaves the file behind without updating `files.json`.  Fix for that is on my TODO list.

As to what to try next, I suggest you try increasing the heap size available to MW by editing the start script, it's the value after the `-Xmx` switch.  Meanwhile I will ask the I2P devs what is causing the leaseset request fails; it shouldn't be related to the number of shared files in any way though.

**Comment by LoveIsGrief** [comment]
> if MW loses the I2CP connection to the I2P router it will shut down

It might be necessary to decouple MuWire from the network connection.

How can I log absolutely everything MuWire does? It might help debugging.

**Comment by zlatinb** [comment]
I checked with the I2P devs and the disconnection is most likely due to MW being hung and not responding to the router over I2CP quickly enough.  So it is definitely an MW issue, not an I2P issue.

To log everything, copy `logs/4_logging.properties` to `logging.properties` and put it in PWD of where the `muwire` script is launched.  The logfile should appear in PWD and will be called MuWire.log.

**Comment by LoveIsGrief** [comment]
I'm definitely getting a lot of closes e,g

```
java.lang.Exception: Input stream closed
        at net.i2p.client.streaming.impl.MessageInputStream.closeReceived(MessageInputStream.java:302)
        at net.i2p.client.streaming.impl.Connection.disconnect(Connection.java:728)
        at net.i2p.client.streaming.impl.Connection.disconnect(Connection.java:702)
        at net.i2p.client.streaming.impl.Connection.waitForConnect(Connection.java:1189)
        at net.i2p.client.streaming.impl.ConnectionManager.connect(ConnectionManager.java:523)
        at net.i2p.client.streaming.impl.I2PSocketManagerFull.connect(I2PSocketManagerFull.java:567)
        at net.i2p.client.streaming.impl.I2PSocketManagerFull.connect(I2PSocketManagerFull.java:627)
```

and

```
java.lang.Exception: cause
        at net.i2p.client.streaming.impl.Connection.disconnect(Connection.java:748)
        at net.i2p.client.streaming.impl.Connection.disconnect(Connection.java:702)
        at net.i2p.client.streaming.impl.Connection.waitForConnect(Connection.java:1189)
        at net.i2p.client.streaming.impl.ConnectionManager.connect(ConnectionManager.java:523)
        at net.i2p.client.streaming.impl.I2PSocketManagerFull.connect(I2PSocketManagerFull.java:567)
        at net.i2p.client.streaming.impl.I2PSocketManagerFull.connect(I2PSocketManagerFull.java:627)
        at com.muwire.core.connection.I2PConnector.connect(I2PConnector.groovy:17)
```

Doesn't seem to be related. I'll try and debug why is just stops hashing but continues writing the `files.json`

**Comment by LoveIsGrief** [comment]
You were right, it's the memory. After increasing it to 1024MB, it now continues hashing, however it does take a pretty long time. The operation that takes the longest time is copying the `files.json` from `tmp` to `home` (~15s).

I'll open a new issue for that since, that isn't optimized

**Comment by LoveIsGrief** [comment]
I'm reopening this because I do believe that loading the entire files database into memory isn't a good idea and requires to much RAM. Either there's a problem with what is stored in the DB (too big) or that everything is kept in RAM.

**Comment by zlatinb** [comment]
While I agree that ideally only parts of the database (if you can call it that way ) should be loaded in RAM, implementing this would be a pretty serious task, especially since in order to respond to a query at least the keyword index (instance of `SearchIndex`) needs to be fully loaded in memory.

It may be easiest to only offload the mapping `<File,SharedFile>` to disk where the `SharedFile` instances are loaded on demand.  This would have repercussions to upload request handling too, but is doable.

An even smaller change would be to remove some of the "cached" fields from the `SharedFile` object.  Those were added to reduce object churn while serializing `files.json`, but with the proposed changes in #37 they may no longer be necessary.

**Comment by LoveIsGrief** [comment]
I'm not sure yet how much different the system is from bittorrent, but bittorrent clients are able to share multi-GB torrents without running out of memory. They must be doing something differently.

> in order to respond to a query at least the keyword index (instance of `SearchIndex`) needs to be fully loaded in memory.

That seems to only include filenames and comments. I'm starting to question if a real database is needed. Searches for comments, filenames, hashes, etc. might  force the move towards that in order to keep a low memory footprint, but a bigger executable. Would that be acceptable?  
#36 has the sizes of the  embedded dbs

**Comment by zlatinb** [comment]
The system is quite different from BitTorrent because of the search capability.  The biggest consumer of memory of the top of my head (i.e. without having profiled) are the hash list arrays which can be up to a few hundred KB per file.  

The `SharedFile` object contains a reference to an `InfoHash` object which contains the hash list array and the hash of that array which is the "root" hash.  I would suggest we look into loading `SharedFile` objects on demand from disk after #37 is merged.  It would be much simpler change which preserves the search behavior; moving to a proper database would require a lot of massaging to preserve that.

**Comment by LoveIsGrief** [comment]
OK :slightly_smiling_face:  
If I have time I'll have a look at how on-demand loading could be done, unless you know you want to take over the task. Then I would find something else to  do :)

**Comment by zlatinb** [comment]
On-demand loading requires that every `SharedFile` object is stored in a separate file.  I'll review #37 carefully and after that is merged on-demand loading should be trivial.

**Comment by LoveIsGrief** [comment]
Wonderful. Thank you for being so responsive and considerate.

**Comment by zlatinb** [comment]
@LoveIsGrief started work in branch "separate-hashlist".  The logic is to store the hashlist in a separate file with extention `.hashlist` in the same place as the `.json` file for each shared file.  

Since the change to folder persister was never released I will not support converting from folder persister to separate hashlist, only from `files.json`.  

Pls test with large library.  Uploads don't work (as of time of this comment) but memory usage should be down significantly.

**Comment by zlatinb** [comment]
And uploads working :)

**Comment by zlatinb** [comment]
Merged to master.  Please test at your convenience

**Comment by LoveIsGrief** [comment]
That was quick 😄
I'll try and test it next week when I have time. 

Have a great weekend!

**Comment by zlatinb** [comment]
Known bug: comments added through the gui (i.e. not through sidecar files) will not persist across restart.

**Comment by LoveIsGrief** [comment]
Hey @zlatinb,

I let it run for a day and was able to index everything :tada: Maybe somebody will download something to make the test complete.
Should I leave this open until you fix the comments?

**Comment by zlatinb** [comment]
Good to hear it hashed everything.  Yes please leave issue open until comments are fixed, there may be other regressions too as this was a very invasive change.

**Comment by zlatinb** [comment]
Comments should be fixed with latest commit.  If I don't hear of any other regressions within a week I'll close the ticket.

**Comment by zlatinb** [comment]
0.6.10 is out with fixed comment persistence, no other regressions have been reported.  Closing this ticket
